### PR TITLE
Solution for hallPeriodCounter conflict

### PR DESCRIPTION
**Pull request by PeterChmurciak**
Different way of calculating RPM that does not rely on custom variable incremented by TIMER. It utilizes only external interrupt, not sure if that one caused the conflict also, but I do not think it should.

### CODE REVIEW

**Comment by PeterChmurciak** [comment]
@gergelytakacs I think this is a good solution to the [issue #204](https://github.com/gergelytakacs/AutomationShield/issues/204#event-3047001241). 

By using different method for calculating the RPM I was able to remove the conflicting parts. 
New method uses the known sampling period and counts the amount of pulses that occur during a set amount of samples - during a known amount of time. Results look very good:
(The parts under 15 and above 245 pwm are not linear so they were left out during fitting)

StaticPWM100_6sample

As this method is as good as the previous one, while being less complicated I think we can use it and return to having both .h and .cpp files. What do you think ?

**Comment by PeterChmurciak** [comment]
@gergelytakacs It seems that I was able to update `vl53l0x-arduino` submodule from my repository that includes modifications required for Arduino-DUE compatibility. 

With those changes in place and using RPM calculating method mentioned above, I was able to run Open-loop example and RPM controlling example (possible only for R2) of FloatShield R2, on all boards that I have at home (UNO,MEGA,ZERO,DUE). I will have to confirm that, but it looks like FloatShield R2 is now fully compatible with all mentioned boards. Its very unfortunate the fan itself is no good...

For illustration the best PID results so far on R2:
PID

**Comment by gergelytakacs** [comment]
So, without going over the code (sorry, time-deficiency)
- Before you used a counter (e.g. TCx) to count those impulses and now you simply use a digital interrupt launching an... interrupt? Because that could work, and I'm actually sorry that I have not suggested that.
- If there are no conflicts, then yes, you are free to go back to the much nicer *.cpp and *.h division of the source.
- Based on your chart (measurement results) the PWM counting method seems completely fine to me. It is a shame that we can't really use that. Maybe you design a much better hardware during your PhD? :)))
- Due compatibility is excellent. Remember, that the R1 wasn't tested with the Due either, so this is a very important step.

File: examples/FloatShield/FloatShield_PID/FloatShield_PID.ino [hunk 1]
```diff
@@ -36,9 +36,15 @@ float u = 0.0;                                                    // Input (Fan
...
-#define TD 0.01           // PID Td constant
+#if SHIELDRELEASE == 1
+  #define KP 0.25           // PID Kp constant
+  #define TI 5              // PID Ti constant
+  #define TD 0.01           // PID Td constant
+#elif SHIELDRELEASE == 2
+  #define KP 0.01           // PID Kp constant
+  #define TI 2              // PID Ti constant
+  #define TD 0.01           // PID Td constant
+#endif
```
Comment (gergelytakacs) [inline examples/FloatShield/FloatShield_PID/FloatShield_PID.ino]: Okay, this is good.

File: src/FloatShield.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+#include "FloatShield.h"         // Include header file
+
+#ifdef VL53L0X_h                 // If library for distance sensor was sucessfully included
+
+#if SHIELDRELEASE == 2           // For hardware release version 2  
+  void hallPeriodCounter(void) {                                  // Interrupt routine on external interrupt pin 1
+    FloatShield.pulseCount++;                                     // Increment the pulse counter
+    FloatShield.pulseMeasured = 1;                                // Set flag that pulse was sucessfully measured
+  }
+#endif
```
Comment (gergelytakacs) [inline src/FloatShield.cpp]: Ah, ok. So this is like an external interrupt, not using the TC.

File: src/FloatShield.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      _rpm = pulseCount * _pulseCountToRPM;               // Calculate RPM
+      pulseCount = 0;                                     // Reset pulse counter
+      sampleCounter = 0;                                  // Reset sample counter 
+    }
+    return _rpm;                                          // Return RPM value
+  } else                                                  // If pulse length was not correctly measured - no power going to the fan
+    pulseCount = 0;
+    sampleCounter = 0;                                     
+  return 0;                                               // Return 0
+}
```
Comment (gergelytakacs) [inline src/FloatShield.cpp]: Haven't really analyzed it line to line, but your tests show that it is good.

**Comment by PeterChmurciak** [comment]
@gergelytakacs To better explain the new counting method:

* By using the external interrupt the amount of pulses is being continuously counted
* Every n-th sample, where "n" is based on the sampling period (Ts) used, the RPM is calculated and the pulse counter is reset
* We calculate RPM from the known amount of pulses that occur in a set period of time (n*Ts) as every pulse in this case represents one quarter of a rotation
* Amount of samples "n" is calculated from used sampling period in a way, to guarantee that the resolution of the measurement will be ~100 RPM - for example let's say we will count the amount of pulses that occur within `n*Ts` seconds, we want the "n" to be such value as to allow that if we count `x` pulses, the RPM will be 1000 and if we count `x+1` pulses, the RPM will be ~1100
* Currently by default the `Ts` is expected to be 25ms and the `n` is 6 samples to allow exactly 100 RPM resolution - every 6 samples (150ms) the RPM is updated. If the user wanted to use different sampling period, he would use `FloatShield.setSamplingPeriod(samplingPeriodInMilliseconds)` - to use different "n" to maintain the ~100 RPM resolution.
* I have tried simply calculating the RPM in every sample - the amount of pulses that occur within one sample, but with sampling period 25ms, the resolution was 600 RPM (if we count `x` pulses, the RPM will be 1000 and if we count `x+1` pulses, the RPM will be 1600)  which I found insufficient, therefore all the fuss with the number of samples "n"

**Pull request merged**